week3/main.cpp

```diff
@@ -0,0 +1,139 @@
+                                TheDictionary, Prefixes, Answer, SamePaths);
+            }
+        }
+        if (index % NumberOfColumns != NumberOfColumns - 1) { // East
+            WordsFromPrefix(index + 1, prefix + Grid[index + 1], NumberOfColumns, NumberOfRows, Grid,
+                            TheDictionary, Prefixes, Answer, SamePaths);
+            if (index / NumberOfColumns != 0) { // North-East
+                WordsFromPrefix(index + 1 - NumberOfColumns, prefix + Grid[index + 1 - NumberOfColumns],
+                                NumberOfColumns, NumberOfRows, Grid, TheDictionary, Prefixes, Answer, SamePaths);
+            }
+            if (index / NumberOfColumns != NumberOfRows - 1) { // South-East
+                WordsFromPrefix(index + 1 + NumberOfColumns, prefix + Grid[index + 1 + NumberOfColumns], NumberOfColumns, NumberOfRows, Grid,
+                                TheDictionary, Prefixes, Answer, SamePaths);
+            }
+        }
+        if (index / NumberOfColumns != 0) { //North
+            WordsFromPrefix(index - NumberOfColumns, prefix + Grid[index - NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Prefixes, Answer, SamePaths);
+        }
+        if (index / NumberOfColumns != NumberOfRows - 1) { // South
+            WordsFromPrefix(index + NumberOfColumns, prefix + Grid[index + NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Prefixes, Answer, SamePaths);
+        }
+    }
+}
+
+
+std::set<std::string> WordsFound(int NumberOfRows, int NumberOfColumns, char* Grid,
+                                 std::set<std::string> &TheDictionary) {
+    // build prefixes
+    std::set<std::string> Prefixes;
+    std::string word1 = "";
+    for(std::set<std::string>::iterator it = TheDictionary.begin(); it != TheDictionary.end(); it++) {
+        word1 = "";
+        for(int i = 0; i < (*it).length(); ++i) {
+            word1 = word1 + (*it)[i];
+            Prefixes.insert(word1);
+        }
+    }
+    // launch search from every cell
+    std::set<std::string> Answer;
+    word1 = "";
+    std::set<std::pair<int, std::string>> SamePaths;
+    for(int i = 0; i < NumberOfColumns*NumberOfRows; ++i) {
+        word1 = Grid[i];
+        WordsFromPrefix(i, word1, NumberOfColumns, NumberOfRows, Grid, TheDictionary, Prefixes, Answer, SamePaths);
+    }
+    return Answer;
+}
+
+
+int main() {
+    // input data
+    std::cout << "Let's make a grid" << std::endl;
+    std::cout << "Enter the number of rows: ";
+    int NumberOfRows;
+    std::cin >> NumberOfRows;
+    std::cout << std::endl << "Enter the number of columns: ";
+    int NumberOfColumns;
+    std::cin >> NumberOfColumns;
+    std::cout << std::endl;
+    std::cout << "Now enter the grid" << std::endl;
+    char* Grid = new char[NumberOfRows * NumberOfColumns];
```

Comment:
Using a 1-d array instead of a 2-d one is a common design choice, but why did you preferred the 1-d array here? This kind of decision is a good thing to have comments about :)